speechbrain/core.py

```diff
@@ -1050,100 +1069,45 @@ def fit_batch(self, batch):
-                    self.optimizer_step += 1
-        else:
-            if self.bfloat16_mix_prec:
-                with torch.autocast(
-                    device_type=torch.device(self.device).type,
-                    dtype=torch.bfloat16,
-                ):
-                    outputs = self.compute_forward(batch, Stage.TRAIN)
-                    loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-            else:
-                outputs = self.compute_forward(batch, Stage.TRAIN)
-                loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
+        amp = AutocastConfig.from_name(self.autocast)
+        should_step = (self.step % self.grad_accumulation_factor) == 0
+
+        with self.no_sync(not should_step):
+            with torch.autocast(
+                dtype=amp.dtype, device_type=torch.device(self.device).type,
+            ):
+                outputs = self.compute_forward(batch, sb.Stage.TRAIN)
 
-            if self.check_gradients(loss):
-                valid_loss = True
-                self.valid_step += 1
+            # The objectives are removed from Autocast to avoid
+            # potential numerical instabilities with CTC loss, etc.
+            loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+            loss = loss / self.grad_accumulation_factor
 
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    (loss / self.grad_accumulation_factor).backward()
+            if self.try_safe_backward(loss):
                 if should_step:
-                    self.optimizer.step()
-                    self.zero_grad()
-                    self.optimizer_step += 1
+                    self.optimizers_step()
 
         self.on_fit_batch_end(batch, outputs, loss, should_step)
         return loss.detach().cpu()
 
-    def on_fit_batch_end(self, batch, outputs, loss, should_step):
-        """Called after ``fit_batch()``, meant for calculating and logging metrics.
-
-        Arguments
-        ---------
-        batch : list of torch.Tensors
-            Batch of data to use for training. Default implementation assumes
-            this batch has two elements: inputs and targets.
-        outputs : list or dictionary of torch.Tensors
-            Returned value of compute_forward().
-        loss : torch.Tensor
-            Returned value of compute_objectives().
-        should_step : boolean
-            Whether optimizer.step() was called or not.
-        """
-        pass
+    def check_gradients(self):
+        """ Checks if the gradients are finite. If not, it will emit a warning and set them to zero."""
+        for param in self.modules.parameters():
+            if param.requires_grad and param.grad is not None:
+                if not torch.isfinite(param.grad).all():
+                    param.grad = torch.zeros_like(param.grad)
```

Comment:
I think part of the original reason for checking the loss instead of the gradient was out of concern for the speed of the operation (under the theory that checking millions/billions of parameters could be slow). Do you know if this operation adds a noticeable amount of time to each step or is it fast?

Reply:
One note: I used `param.grad = None` instead of `param.grad = torch.zeros_like(param.grad)` (should be faster). And also, I don't know if you saw it, but NeMO is doing something a little bit different see: https://github.com/NVIDIA/NeMo/blob/1c10b3677040cde4785ea73f7d9ec887578885c7/nemo/collections/asr/models/asr_model.py#L120. What do you think about how they are handling NaN/infs gradients?

Reply:
Btw, ce could turn `check_gradients` optional. We could add a new flag e.g., `skip_grad_nans` in SB to activate the check+replacement of grads.

Reply:
While 0.04s per iteration doesn't sound like a lot, perhaps it would still make sense to turn `check_gradients` to be optional just in case there's a weird scenario that it matters a lot.

Reply:
Yep did it.